README.md

```diff
@@ -32,165 +83,450 @@ PIG generates the following components in Kotlin (and may generate similar compo
+            (lit value::int)
+            (binary op::operator left::expr right::expr)
+        )
+    )
+)
+```
 
-Note that the `toy_lang.build` is not generated yet but will be soon.
+This involves two `sum` types: `operator` and `expr`.
+
+The `operator` `sum` declares five different arithmetic operations: `plus`, `minus`, `times`, `divide`, and `modulo`.
+Due to the syntax used here (more on that later), each of these is a `product` type, and each gets a Kotlin class
+similar to the example shown above (shown below). However, none of these have any elements.
+
+The `expr` `sum` defines two possible types of expressions that exist our toy calculator:
+
+- `lit`
+- `binary`
+
+The Kotlin equivalent of a `sum` type is a [`sealed class`](https://kotlinlang.org/docs/sealed-classes.html).
 
 ```Kotlin
-val onePlusOne = toy_lang.build { plus(lit(ionInt(1)), lit(ionInt(1))) }
-
-val defineFunc = toy_lang.build { 
-    let(
-        "someFunc", 
-        function(
-            "someArg", 
-            plus(
-                variable("someArg"), 
-                lit(1))),
-        call(
-            variable("someFunc"), 
-            lit(1)))
+class CalculatorAst private constructor() {
+    // ... other stuff 
+    sealed class Expr(override val metas: MetaContainer = emptyMetaContainer()) : CalculatorAstNode() {
+        class Lit(
+            val value: LongPrimitive,
+            override val metas: MetaContainer = emptyMetaContainer()
+        ) : Expr() {
+            // ... API same as "person" type from a previous example
+        }
+
+        class Binary(
+            val op: Operator,
+            val left: Expr,
+            val right: Expr,
+            override val metas: MetaContainer = emptyMetaContainer()
+        ) : Expr() {
+            // ... API same as "person" type from a previous example
+        }
+    }
+    // ... other stuff
 }
 ```
 
-#### Domain Type Instance -> `IonElement` Transformation
+The builder API shown in previous examples works here as well:
+
+```Kolin
+// 1 + 2 * 3
+val expr = CalculatorAst.build {
+    binary(
+        plus(),
+        lit(1),
+        binary(
+            times(),
+            lit(2),
+            lit(3)))
+}   
+```
+
+#### Converter\<T\>
+
+PIG generates a facility that allows an instance of a `sum` type to be converted to any other type.
 
-Transformation to `IonElement` is accomplished by invoking the `.toIonElement()` node which is present on every 
-generated type:
+Two examples are provided below.
 
 ```Kotlin
-println(onePlusOne.toIonElement().toIonText())
-// prints:  (+ (lit 1) (lit 1))
+val expr = CalculatorAst.build {
+    // 1 + 2 * 3 (from the example above)
+}
+
+/** Evaluates an instance of CalculatorAst.Expr */
+class CalculatorAstEvaluator : CalculatorAst.Expr.Converter<Long> {
+    override fun convertLit(node: CalculatorAst.Expr.Lit): Long = node.value.value
+    override fun convertBinary(node: CalculatorAst.Expr.Binary): Long =
+        when (node.op) {
```

Comment:
both of these examples could be refactored a bit by walking the tree once before the `when` instead of doing that in every branch.